mailpoet/lib/Tasks/Sending.php

```diff
@@ -208,18 +221,37 @@ public function taskSubscribers() {
   }
 
   public function getSubscribers($processed = null) {
-    $subscribers = $this->taskSubscribers->getSubscribers();
-    if (!is_null($processed)) {
-      $status = ($processed) ? ScheduledTaskSubscriber::STATUS_PROCESSED : ScheduledTaskSubscriber::STATUS_UNPROCESSED;
-      $subscribers->where('processed', $status);
+    if (is_null($processed)) {
+      $subscribers = $this->scheduledTaskSubscribersRepository->findBy(['task' => $this->task->id]);
+    } else if ($processed) {
+      $subscribers = $this->scheduledTaskSubscribersRepository->findBy(
+        ['task' => $this->task->id, 'processed' => ScheduledTaskSubscriberEntity::STATUS_PROCESSED]
+      );
+    } else {
+      $subscribers = $this->scheduledTaskSubscribersRepository->findBy(
+        ['task' => $this->task->id, 'processed' => ScheduledTaskSubscriberEntity::STATUS_UNPROCESSED]
+      );
     }
-    $subscribers = $subscribers->findArray();
-    return array_column($subscribers, 'subscriber_id');
+
+    return array_map(
+      function(ScheduledTaskSubscriberEntity $scheduledTaskSubscriber) {
+        $subscriber = $scheduledTaskSubscriber->getSubscriber();
+
+        if ($subscriber instanceof SubscriberEntity) {
+          return (string)$subscriber->getId();
```

Comment:
Calling `$subscriber->getId()` to my surprise means that Doctrine is performing an additional query against `wp_mailpoet_subscribers` for every subscriber. This should not be necessary as the subscriber ID is available in `wp_mailpoet_scheduled_task_subscribers`. I was not able to find an easy way to get the subscriber ID using doctrine without making it fetch the full record from `wp_mailpoet_subscribers`. Any ideas on how to achieve that?

I haven't checked but maybe this is happening because of `SafeToOneAssociationLoadTrait` that is used in `ScheduledTaskSubscriberEntity`?

Reply:
We discussed this when we met in London, and if I remember correctly, the mentioned trait causes this loading of each subscriber entity. The proposed solution can be a getter that doesn't use this trait or avoid using the trait because it's not needed anymore. Is it correct @rodrigoprimo?

Reply:
@lysyjan, indeed we discussed this in London. I implemented a getter that doesn't use the trait here: https://github.com/mailpoet/mailpoet/pull/5018/commits/4703a106a0fa83b5c596b5b1f4576e202542e91a#diff-27ecf3a634794e8f2882e6d154694bd3dde08a5ba96882ba0d7dc5e82452c576R130. I don't recall anyone mentioning that stop using the trait as it is not needed anymore was a possibility so I didn't consider this option.

I'm just waiting for the CircleCI build to finish to assign this PR for a final review.